### ISSUE TITLE
Axios Timeout behaviour

### ISSUE THREAD

**Issue by adrys-lab** [JustinBeckwith/retry-axios#138]
We did not find any information about how the timeout acts when retry behaviour comes up.

given

```
axios.post(path, body, {
    baseURL: ur;,
    timeout: 12000,
    raxConfig: {
      retry: 5,
      retryDelay: 500,
    },
```

is the timeout of 12sec apply for all the 5 retries ? so 12 sec for performing 5 possible requests ?
or for each retry, the timeout is reset to 12sec ? being possibly 12sec * 5 attempts ?

can you please introduce this documentation in the main README ?

**Comment by anmol098** [comment]
@adrys-lab
Use 
backoff : 'static'
This will give static delay between retry. 
For the snippet you pasted. 
It will retry after 500ms

**Comment by adrys-lab** [comment]
@anmol098 thanks for the backoff topic,

but i was talking about the general HTTP Timeout request when the server-side takes more than 12000 to respond.

My question is, if its a timeout for each retry, or by contrast, for ALL retries together.

**Comment by adrys-lab** [comment]
Is there a way to make the `axios` timeout NOT BEING RESET to 0 for each retry ?

if we setup 10 seconds of timeout, 
we want to interpret it as a global value, 
so it is not used for each retry but for the whole request lifecycle.

**Comment by JustinBeckwith** [comment]
Yeah, there's no way to do that directly through `retry-axios` right now.  The `timeout` is passed directly to Axios, and taken into account for each retry.  You'd need to tune the `retries` variable to match the number you're looking for, otherwise you're going to get your `timeout` and the retry logic fighting with each other.

**Comment by patryk-smc** [comment]
Hey everyone,
I think I have a similar problem. I'm dealing with an API that has a tendency to timeout. It retries after first timeout, and if the subsequent request timeouts too - it fails. I set `noResponseRetries` to 3, so I would expect that it would retry three times, not just once. What I'm missing here?

```
import * as rax from "retry-axios";
import axios from "axios";

export default function callApi() {
  const client = axios.create({
    baseURL: "https://example.com",
    timeout: 1000,
  });

  rax.attach(client);

  return new Promise((resolve, reject) => {
    client({
      method: "get",
      url: "/api",
      params: {},
      raxConfig: {
        retry: 3,
        noResponseRetries: 3,
        backoffType: "static",
        onRetryAttempt: (err) => {
          const cfg = rax.getConfig(err);
          console.log(`Retry attempt #${cfg.currentRetryAttempt}`);
        },
      },
    })
      .then((r) => {
        if (r.data) {
          resolve(r.data);
        } else {
          resolve(null);
        }
      })
      .catch((e) => {
        console.error(e);
        reject(e);
      });
  });
}
```

**Comment by jmurty** [comment]
Please consider adding a feature to restrict retries to the original request timeout. The competing [axios-retry](https://github.com/softonic/axios-retry#options) library has something along these lines under the `shouldResetTimeout` option.

In the meantime, I have implemented something workable but kludgy via a `retriesStopAtTimeout` option with following code, which I'll leave here in case it helps someone else:
```javascript
const retryAxios = require("retry-axios");

const axiosClient = axios.create(config);

axiosClient.defaults.raxConfig = {
  instance: axiosClient,

  // Apply your default retry settings here..

  /**
   * HACK: Add and enforce option `retriesStopAtTimeout` which will prevent
   * retry requests after the original Axios timeout. For example, if the
   * timout was 2000, retry requests will only be sent within that 2 second
   * timeframe (though the final request may finish after that time)
   */
  retriesStopAtTimeout: true,
  shouldRetry: function enforceRetriesStopAtTimeout(err) {
    const raxConfig = retryAxios.getConfig(err);

    // Ensure max retries is always respected
    if (raxConfig.currentRetryAttempt >= raxConfig.retry) return false;

    // Respect 'retry-axios' decision not to retry
    if (!retryAxios.shouldRetryRequest(err)) {
      return false;
    }

    // Only abort retries at timeout cutoff if configured to do so
    if (!raxConfig.retriesStopAtTimeout) {
      return true;
    }

    // This delay calculation is adapted from 'retry-axios' index.ts
    let delay;
    if (raxConfig.backoffType === "linear") {
      delay = raxConfig.currentRetryAttempt * 1000;
    } else if (raxConfig.backoffType === "static") {
      delay = raxConfig.retryDelay;
    } else {
      delay = ((Math.pow(2, raxConfig.currentRetryAttempt) - 1) / 2) * 1000;
    }

    // Will the next retry request be sent before the retries cutoff?
    raxConfig.nextRetryAvailableTime =
      raxConfig.retriesStopAtTimeoutCutoff - Date.now() - delay;

    // Don't retry request if there is no time for it to complete
    if (raxConfig.nextRetryAvailableTime <= 0) {
      return false;
    }

    return true;
  },
};
retryAxios.attach(axiosClient);

/**
 * HACK: Store the cutoff time to apply if `retriesStopAtTimeout` is set
 * when the first request is sent.
 */
axiosClient.interceptors.request.use(function storeRetriesStopAtTimeoutCutoff(config) {
  const raxConfig = config.raxConfig || {};
  if (!raxConfig.retriesStopAtTimeoutCutoff) {
    raxConfig.retriesStopAtTimeoutCutoff = Date.now() + config.timeout;
  }
  return config;
},
undefined);
```

**Comment by kvv002** [comment]
Hi @JustinBeckwith , any update on this enhancement?

**Comment by JustinBeckwith** [comment]
## Current Behavior

Currently, retry-axios **resets the timeout for each retry**. So with your config:

```js
timeout: 12000,  // 12 seconds
retry: 5
```

Each of the 5 retry attempts gets a fresh 12-second timeout, meaning the total time could be up to **60 seconds** (12s × 5 attempts).

## Comparison with axios-retry

The competing library `axios-retry` has a `shouldResetTimeout` option:
- **Default (`shouldResetTimeout: false`)**: Timeout is global across all retries (what you're asking for)
- **`shouldResetTimeout: true`**: Timeout resets per retry (current retry-axios behavior)

## Workaround

As @jmurty showed in their comment, you can implement this using a custom `shouldRetry` function that tracks elapsed time and prevents retries after the original timeout would have expired.

## Recommendation for Future Enhancement

For a future version, we could add a `shouldResetTimeout` configuration option (defaulting to `true` to maintain current behavior) that would allow users to opt into global timeout behavior.

Would that be useful for your use case?

## Documentation

In the meantime, I'll add clarification to the README that:
1. The `timeout` option applies to each individual retry attempt
2. Total execution time = timeout × number of attempts (plus retry delays)
3. Users who want a global timeout should use a custom `shouldRetry` function

**Comment by JustinBeckwith** [comment]
The `timeout` setting in axios applies **per individual retry attempt**, not across all retries.

With your configuration:
```javascript
axios.post(path, body, {
    baseURL: url,
    timeout: 12000,
    raxConfig: {
      retry: 5,
      retryDelay: 500,
    },
})
```

Each of the 5 attempts can take up to 12 seconds. The timeout resets for each retry, so the total possible time is:
- **Up to 60 seconds** (12 sec × 5 attempts) for the requests themselves
- **Plus** the backoff delays between retries

The `timeout` is an axios configuration option that retry-axios doesn't modify. When a retry occurs, axios uses the same config (including the timeout) for each subsequent attempt.

So to answer your question: the timeout is reset to 12 seconds for each retry attempt, not 12 seconds total for all 5 attempts.
